Refactor query parsing

shubhankarunhale opened the pull request:
Added dictionaries for queries in each function.

ttt-77 commented:
Can you change it to the format {'test_query_i' : {'query_str':, 'normalized_query_str':, 'correct_service':, 'correct_tables':}}?

shubhankarunhale commented:
@ttt-77 Okay sure, updating the dictionary format. Also, please let me know if I should add a key with the num_query parameter as well in the inner dictionary.

shubhankarunhale commented:
Hi @ttt-77 , I have updated the dictionary structure as per your request. Also added an additional key "num_query" to each query dictionary to enable iteration of the test queries. Can you please review?

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -40,174 +40,161 @@ async def test_nested_query(self):
+    
+    queries = {
+    "test_query_1":{
+        "query_str": '''
```
Better to define constant parameters for these variables, like QUERY_STR = 'query_str'

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -40,174 +40,161 @@ async def test_nested_query(self):
-
+    
+    queries = {
+    "test_query_1":{
```
start with 0

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -40,174 +40,161 @@ async def test_nested_query(self):
+        "num_query":1
+    }
+}
+    for i in range(1,10):
```
use 'for i in range(len(queries)):'

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -40,174 +40,161 @@ async def test_nested_query(self):
+    }
+}
+    for i in range(1,10):
+      self._test_query(queries["test_query_"+str(i)]["query_str"], config, queries["test_query_"+str(i)]["normalized_query_str"], queries["test_query_"+str(i)]["correct_fp"],queries["test_query_"+str(i)]["correct_service"], queries["test_query_"+str(i)]["correct_tables"], queries["test_query_"+str(i)]["num_query"])
```
refactor '_test_query' to receive only two parameters "queries[f'test_query_{i}']" and config

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -252,44 +238,45 @@ async def test_correct_approximate_query(self):
+    
+    queries = {
+    # approx aggregation as a subquery
+    "test_query_1":{
```
Same modification logic as above

ttt-77 commented:
Can you also help refactor 'test_udf_query' with the similar format? @shubhankarunhale

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -11,25 +11,33 @@
 
 DB_URL = "sqlite+aiosqlite://"
 
+QUERY_STR = "query_str"
```
single quotation

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -40,174 +48,162 @@ async def test_nested_query(self):
+    }
+}
+    for i in range(len(queries)):
+      self._test_query(queries["test_query_"+str(i)], config)
```
use "queries[f'test_query_{i}']"

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -252,44 +247,45 @@ async def test_correct_approximate_query(self):
+    }
+}
+    for i in range(len(queries)):
+      self._test_query(queries["test_query_"+str(i)], config)
```
same as above

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -11,25 +11,40 @@
 
 DB_URL = "sqlite+aiosqlite://"
 
+QUERY_STR = 'query_str'
```
Add comments to separate different parts' parameters.Something like 'normal query extraction parameters', 'udf query extraction parameters'.

shubhankarunhale commented:
Hi @ttt-77 , I have implemented all the review comments. Can you please review?

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -11,25 +11,43 @@
 
 DB_URL = "sqlite+aiosqlite://"
 
+#normal query extraction parameters
```
space between '#' and the character

ttt-77 commented:
Others looks good to me.

shubhankarunhale commented:
@ttt-77 , Okay great! I have added the space to the comments.

ttt-77 commented:
Have you merged the newest code from main branch? The test didn't pass.

shubhankarunhale commented:
> Have you merged the newest code from main branch? The test didn't pass.

Yes, I have merged the latest code into my branch.  
The test is failing at 
`dataframe_sql, query_after_extraction = query.udf_query`. with a KeyError
I am rechecking my refactored code, and looking into the failure.

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -325,426 +330,427 @@ def _test_equality(test_query_list, config):
-
+    },
+    # test user defined function for exact aggregation query
+    "test_query_7" : {
```
Remove this one

shubhankarunhale commented:
@ttt-77 I have removed the the exact aggregation query

shubhankarunhale commented:
@ttt-77  thanks for the input, the tests are passing now. Can you please take a look?

ttt-77 commented:
@ddkang Could you please review this PR?

ddkang commented on tests/test_query_parsing.py:
```diff
@@ -11,25 +11,43 @@
+# normal query extraction parameters
+QUERY_STR = 'query_str'
+NORMALIZED_QUERY_STR = 'normalized_query_str'
+CORRECT_FP='correct_fp'
```
spaces after =

ddkang commented on tests/test_query_parsing.py:
```diff
@@ -219,27 +225,26 @@ async def test_approximate_query_as_subquery(self):
-    query_str1 = '''
+    queries = {
+      # approx aggregation as a subquery
+      "query_str1":'''
```
spaces

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -11,25 +11,43 @@
+# normal query extraction parameters
+QUERY_STR = 'query_str'
+NORMALIZED_QUERY_STR = 'normalized_query_str'
+CORRECT_FP= 'correct_fp'
```
space before =

shubhankarunhale commented:
> @shubhankarunhale please tag the issue in development if this resolves an issue
@ddkang  I have added the reference to issue

ddkang commented on tests/test_query_parsing.py:
```diff
@@ -252,60 +257,60 @@ async def test_correct_approximate_query(self):
+        # filter predicates connected by OR are in same set
+        CORRECT_SERVICE: [{}],
+        CORRECT_TABLES: [{}],
+        NUM_QUERY: 2
```
Why is this hard coded?

ttt-77 replied:
This is used to check the correctness of parsing results. The hard coded parts are desired results.

ddkang replied:
What does NUM_QUERY: 2 mean?

ttt-77 replied:
The query contain a subquery, so there are two queries in total. Or the number of 'SELECT' in the query sentence.

ddkang replied:
This should probably be renamed to something like NUMBER_OF_SELECT_CLAUSES or something shorter

shubhankarunhale replied:
Okay sure, renaming it

ddkang commented on tests/test_query_parsing.py:
```diff
@@ -11,25 +11,43 @@
 DB_URL = "sqlite+aiosqlite://"
 
+# normal query extraction parameters
+QUERY_STR = 'query_str'
```
This entire thing should probably be an Enumerate

shubhankarunhale replied:
Okay, sure.  Changing to an Enum

shubhankarunhale replied:
@ddkang Done with the suggested changes

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -8,28 +8,51 @@
+  CORRECT_TABLES = 'correct_tables'
+  NUM_OF_SELECT_CLAUSES = 'num_of_select_clauses'
+
+  # udf query extraction parameters
```
Use another class for udf query

shubhankarunhale replied:
Done, added another class.

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -8,28 +8,55 @@
+  def __str__(self) -> str:
+        return self.value
+
+class UDFQueryParameters(str, Enum):
```
UdfQueryParameters

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -40,174 +67,162 @@ async def test_nested_query(self):
+       
+    queries = {
+    "test_query_0": {
+        QueryParameters.QUERY_STR: '''
```
Can you reformat here and below? Start with a new line with four space like
QueryParameters.QUERY_STR:
        '''
        SELECT
        FROM
        '''

shubhankarunhale replied:
@ttt-77 Can you please let me know if I should reformat all the other fields (CORRECT_FP,etc) in this format as well, or just the query and normalized query fields?

ttt-77 replied:
All

shubhankarunhale replied:
Done, reformatted all the dictionaries.

ttt-77 commented:
Looks good to me. Could you please review this PR? @ddkang

ddkang commented on tests/test_query_parsing.py:
```diff
@@ -8,28 +8,55 @@
 from tests.inference_service_utils.inference_service_setup import \
     register_inference_services
 from tests.utils import setup_gt_and_aidb_engine
+from enum import Enum
```
isort

ddkang commented on tests/test_query_parsing.py:
```diff
@@ -40,174 +67,220 @@ async def test_nested_query(self):
+       
+    queries = {
+    "test_query_0": {
+        QueryParameters.QUERY_STR:
```
Why aren't these dataclasses instead of dicts actually?

shubhankarunhale replied:
I see no issues with moving to dataclasses, working on the changes.

shubhankarunhale commented:
Done with the following changes:

- Replaced dictionaries with dataclasses in queries
- Removed constant enums as the query parameter constants are no longer redundant
- Refactored _test_query and test_udf_query methods to work with dataclasses
- Reformatted all the dataclass queries as per the previously mentioned format

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -11,25 +12,57 @@
+# normal query dataclass
+@dataclass
+class TestQuery():
+    query_str: str
```
two spaces

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -11,25 +12,57 @@
+    correct_service: list
+    correct_tables: list
+    num_of_select_clauses: int
+
```
two blank lines

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -40,174 +73,164 @@ async def test_nested_query(self):
+         "WHERE colors02.object_id > (SELECT AVG(object_id) FROM objects00 WHERE frame > "
+         "(SELECT AVG(frame) FROM blobs_00 WHERE frame > 500))"),
+        # test replacing column with root column in filtering predicate,
+        # the root column of 'colors02.object_id' is 'objects00'
```
remove

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -40,174 +73,164 @@ async def test_nested_query(self):
+        # the root column of 'colors02.object_id' is 'objects00'
+        [["objects00.object_id > (SELECT AVG(object_id) FROM objects00 " 
+          "WHERE frame > (SELECT AVG(frame) FROM blobs_00 WHERE frame > 500))"]],
+        # filter predicates connected by AND are in different set 
```
remove

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -219,27 +242,30 @@ async def test_approximate_query_as_subquery(self):
-                 '''
-
-    parsed_query = Query(query_str2, config)
+    parsed_query = Query(queries["query_str2"], config)
```
merge it with Query(queries["query_str1"], config)

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -11,25 +12,57 @@
 
-  def _test_query(self, query_str, config, normalized_query, correct_fp, correct_service, correct_tables, num_query):
-    query = Query(query_str, config)
+  def _test_query(self, test_query, config):
```
can you move this function and 'are_lists_equal' into dataclass?

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -252,60 +278,61 @@ async def test_correct_approximate_query(self):
-          filter_predicate = filter_predicate.sql()
-
-        assert filter_predicate == test_query['dataframe_sql']['filter_predicate']
+    def _test_equality(test_query, config):
```
can you move this function into dataclass?

shubhankarunhale commented:
@ttt-77 Done with the suggested changes, can you please review?

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -11,26 +12,79 @@
-    query = Query(query_str, config)
+      
+      
+  def _test_query(self, test_query, config):
```
only receive the parameter 'config', modify 'test_query' to 'self' below

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -11,26 +12,79 @@
+  num_of_select_clauses: int
+  
+    
+  def _test_equality(self,test_query, config):
```
same as above

shubhankarunhale commented:
Done with the updates

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -40,174 +94,160 @@ async def test_nested_query(self):
+    )
+}
+    for i in range(len(queries)):
+      queries[f'test_query_{i}']._test_query(queries[f'test_query_{i}'], config)
```
only config is needed

shubhankarunhale replied:
Sorry missed that one, removing it

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -219,30 +259,29 @@ async def test_approximate_query_as_subquery(self):
+          AND table2.object_id > (SELECT AVG(ob.object_id) FROM objects00 AS ob WHERE frame > 500);
+          '''
+    }
+    parsed_query = Query(queries["query_str1"]+queries["query_str2"], config)
```
why you add these two string together?

shubhankarunhale replied:
The intent was to test to merge them into one check as the assert is raised even if one of them does not contain approximate query as a subquery. Please let me know if we need a loop instead.

ttt-77 replied:
use a loop

shubhankarunhale replied:
Done, used a loop.

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -278,9 +278,10 @@ async def test_approximate_query_as_subquery(self):
-    with self.assertRaises(Exception):
-      _ = parsed_query.all_queries_in_expressions
+    for i in range(len(queries)):
+        parsed_query = Query(queries[f'query_str{i}'], config)
```
two spaces for in-block code

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -278,9 +278,10 @@ async def test_approximate_query_as_subquery(self):
+    for i in range(len(queries)):
+        parsed_query = Query(queries[f'query_str{i}'], config)
+        with self.assertRaises(Exception):
+        _ = parsed_query.all_queries_in_expressions
```
Indentation

shubhankarunhale replied:
Done

ttt-77 commented:
@shubhankarunhale Check the running error.

shubhankarunhale commented:
> @shubhankarunhale Check the running error.

Looking into it

shubhankarunhale commented:
Done with fixes

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -11,26 +12,76 @@
-class QueryParsingTests(IsolatedAsyncioTestCase):
+# normal query dataclass
+@dataclass
+class TestQuery(IsolatedAsyncioTestCase):
```
class TestQuery:

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -11,26 +12,76 @@
+      
+      
+  def _test_query(self, config):
+    query = Query((self.query_str), config)
```
query = Query(self.query_str, config)

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -11,26 +12,76 @@
-    assert len(query.all_queries_in_expressions) == num_query
-    self.assertEqual(query.query_after_normalizing.sql_str, normalized_query)
+    assert len(query.all_queries_in_expressions) == self.num_of_select_clauses
+    self.assertEqual(query.query_after_normalizing.sql_str, self.normalized_query_str)
```
assert query.query_after_normalizing.sql_str == self.normalized_query_str

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -11,26 +12,76 @@
-    self.are_lists_equal(query.tables_in_filtering_predicates, correct_tables)
-
+    self.are_lists_equal(and_fp, self.correct_fp)
+    self.are_lists_equal(query.inference_engines_required_for_filtering_predicates, test_query.correct_service)
```
self.are_lists_equal(query.inference_engines_required_for_filtering_predicates, self.correct_service)

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -11,26 +12,76 @@
+
+# udf query dataclasses       
+@dataclass
+class UdfMapping():
```
class UdfMapping:

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -11,26 +12,76 @@
+
+
+@dataclass
+class DataframeSql():
```
class DataframeSql:

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -11,26 +12,76 @@
+
+
+@dataclass
+class UdfTestQuery(IsolatedAsyncioTestCase):
```
class UdfTestQuery:

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -11,26 +12,76 @@
+  def _test_equality(self, config):
+    query = Query(self.query_str, config)
+    dataframe_sql, query_after_extraction = query.udf_query
+    self.assertEqual(query_after_extraction.sql_str, self.query_after_extraction)
```
assert query_after_extraction.sql_str == self.query_after_extraction

ttt-77 commented on tests/test_query_parsing.py:
```diff
@@ -323,428 +345,400 @@ def _test_equality(test_query_list, config):
+        DataframeSql(
+           [UdfMapping(
+               ['col__0', 'col__1'],
+               'max_function'
```
'max_function',

ttt-77 commented:
@shubhankarunhale Before requesting a review, please run tests on your local computer to ensure successful execution.

shubhankarunhale commented:
Hi @ttt-77 , okay sure. I am actually facing an issue with sqlite on my local env, I'll focus on resolving that before any additional commits.

shubhankarunhale commented:
@ttt-77 Done with fixes, local run is passing as well. Can you please review?

ttt-77 commented:
Looks good to me. Could you please review this PR? @ddkang

ddkang commented on tests/test_query_parsing.py:
```diff
@@ -323,428 +345,400 @@ def _test_equality(test_query_list, config):
     config = aidb_engine._config
-
-    test_query_list = []
+    
```
spaces

ddkang commented on tests/test_query_parsing.py:
```diff
@@ -323,428 +345,400 @@ def _test_equality(test_query_list, config):
+            '(function__0 = TRUE) AND (function__1 = TRUE)'
+        )
+    ),
+        
```
spaces

ddkang commented on tests/test_query_parsing.py:
```diff
@@ -323,428 +345,400 @@ def _test_equality(test_query_list, config):
+            "(function__0 = TRUE) AND (function__1 = TRUE) AND (function__2 > function__3)"
+        )
+    ),
+    
```
spaces
